lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4077,35 +4118,117 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     auto &original = getOriginal();
     auto &adjoint = getAdjoint();
     auto adjLoc = getAdjoint().getLocation();
-    LLVM_DEBUG(getADDebugStream() << "Running AdjointGen on\n" << original);
-
-    // Create adjoint blocks and arguments.
-    {
-      PostDominanceOrder postDomOrder(&*original.findReturnBB(), postDomInfo);
-      while (auto *origBB = postDomOrder.getNext()) {
-        auto *adjointBB = adjoint.createBasicBlock();
-        adjointBBMap.insert({origBB, adjointBB});
-        postDomOrder.pushChildren(origBB);
-        // If adjoint block is the adjoint entry, continue.
-        if (adjointBB->isEntry())
-          continue;
-        // Otherwise, add a pullback struct argument to the adjoint block.
-        // TODO: Add bb arguments for adjoint values of active values in the
-        // original dominator block. Propagate adjoint values in adjoint bb
-        // terminators below.
-        auto pbStructLoweredType =
-            getPullbackInfo().getPullbackStructLoweredType(origBB);
-        adjointBB->createPhiArgument(
+    LLVM_DEBUG(getADDebugStream() << "Running AdjointEmitter on\n" << original);
+
+    auto *adjGenEnv = getAdjoint().getGenericEnvironment();
+    auto adjGenSig = adjGenEnv
+        ? adjGenEnv->getGenericSignature()->getCanonicalSignature()
+        : nullptr;
+    Lowering::GenericContextScope genericContextScope(
+        getContext().getTypeConverter(), adjGenSig);
+
+    // Get dominated active values in original blocks.
+    // Adjoint values of dominated active values are passed as adjoint block
+    // arguments.
+    auto *domAnalysis =
+        getContext().getPassManager().getAnalysis<DominanceAnalysis>();
+    auto *domInfo = domAnalysis->get(&original);
```

Comment:
Todo: make `domInfo` a member of `AdjointEmitter`, just like `postOrderInfo`.
Will commit after CI passes.